src/System.Management.Automation/utils/EncodingUtils.cs

```diff
@@ -0,0 +1,136 @@
+            };
+
+        internal static Dictionary<string, Encoding> encodingMap = new Dictionary<string,Encoding>(StringComparer.OrdinalIgnoreCase)
+        {
+            { Ascii, System.Text.Encoding.ASCII },
+            { BigEndianUnicode, System.Text.Encoding.BigEndianUnicode },
+            { Default, ClrFacade.GetDefaultEncoding() },
+            { OEM, ClrFacade.GetOEMEncoding() },
+            { Unicode, System.Text.Encoding.Unicode },
+            { Utf7, System.Text.Encoding.UTF7 },
+            { Utf8, ClrFacade.GetDefaultEncoding() },
+            { Utf8Bom, System.Text.Encoding.UTF8 },
+            { Utf8NoBom, ClrFacade.GetDefaultEncoding() },
+            { Utf32, System.Text.Encoding.UTF32 },
+            { String, System.Text.Encoding.Unicode },
+            { Unknown, System.Text.Encoding.Unicode },
+        };
+
+        /// <summary>
+        /// retrieve the encoding parameter from the command line
+        /// it throws if the encoding does not match the known ones
+        /// </summary>
+        /// <returns>a System.Text.Encoding object (null if no encoding specified)</returns>
+        internal static Encoding Convert(Cmdlet cmdlet, string encoding)
+        {
+            if (string.IsNullOrEmpty(encoding))
+            {
+                // no parameter passed, default to UTF8
+                return ClrFacade.GetDefaultEncoding();
+            }
+
+            Encoding foundEncoding;
+            if (encodingMap.TryGetValue(encoding, out foundEncoding))
+            {
+                return foundEncoding;
+            }
+
+            // error condition: unknown encoding value
+            string validEncodingValues = string.Join(", ", TabCompletionResults);
+            string msg = StringUtil.Format(PathUtilsStrings.OutFile_WriteToFileEncodingUnknown, encoding, validEncodingValues);
+
+            ErrorRecord errorRecord = new ErrorRecord(
+                PSTraceSource.NewArgumentException("Encoding"),
+                "WriteToFileEncodingUnknown",
+                ErrorCategory.InvalidArgument,
+                null);
+
+            errorRecord.ErrorDetails = new ErrorDetails(msg);
+            cmdlet.ThrowTerminatingError(errorRecord);
+
+            return null;
+        }
+
+    }
+
+    /// <summary>
+    /// To make it easier to specify -Encoding parameter, we add an ArgumentTransformationAttribute here.
+    /// When the input data is of type string and is valid to be converted to System.Text.Encoding, we do
+    /// the conversion and return the converted value. Otherwise, we just return the input data.
+    /// </summary>
+    internal sealed class ArgumentToEncodingTransformationAttribute : ArgumentTransformationAttribute
+    {
+        public override object Transform(EngineIntrinsics engineIntrinsics, object inputData)
+        {
+            string encodingName;
+            if (LanguagePrimitives.TryConvertTo<string>(inputData, out encodingName))
```

Comment:
If the `inputData` is already an Encoding object, shouldn't we just let it pass by?

Reply:
And do we really need to convert `inputData` to `string` here? Why not just use `string encodingName = inputData as string`?